### PR TITLE
Pre-commit and post-commit hooks in AbstractKafkaMirrorMakerConnectorTask

### PR DESCRIPTION
**Pull request by jzakaryan**
Sometimes it's necessary to perform pre-commit or post-commit operations in BMM connectors tasks. One could override/extend abstract method `maybeCommitOffsets()` for that purpose, but that method is called each time messages are polled and sent, and not always a call to `maybeCommitOffsets` results in offsets actually being committed.
The changes in this pull request introduce dedicated pre-commit and post-commit hooks that are called before and after a commit attempt. The default implementation of `preCommitHook()` is a no-op, while the default implementation of `postCommitHook()` logs an error if the commit failed and throws an exception. Please note that the latter behavior was previously part of the `commitWithRetries()` method.

Update: Post-commit hook has been removed after discussing with @vmaheshw and @somandal. There's no use case for it now and it can be added back when there is one.

### CODE REVIEW

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java
Comment (abhishekmendhekar) [inline datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:782]: It'll be good if the `postCommitHook()` can get hold of the exception causing the commit to fail.

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java
Comment (abhishekmendhekar) [inline datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:783]: I think the `postCommitHook()` should not throw the exception because if the method is overridden the derived class and it fails to implement this logic then ab exception will never be thrown.
Instead the logic should remain the same as before and this operation can remain as a no-op.

**Review by abhishekmendhekar — COMMENTED** [review]
Since I am not the an approver here just left my thoughts about the code. :)

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java
Comment (somandal) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:783]: +1

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java
Comment (somandal) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:782]: Since we have multiple retries for commit, caching  the last exception should be sufficient.

Comment (jzakaryan) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:782]: I'm not sure if there's value in propagating the Kafka exception up except for logging purposes, and the code in `commitWithRetries` already logs the exception. Do you think that there's value in propagating the last exception, @somandal?

Comment (somandal) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:782]: Right, at the moment I don't really fully see a need for the postCommitHook either, so i'm not yet clear about how the exception will help as such. I'll leave it up to you @jzakaryan to make a call on this.

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java
Comment (somandal) [inline datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:624]: I think there are many other exceptions that can be thrown here, other than KafkaException. (check documentation for commitSync())

I'd suggest adding a catch block for those exceptions too, where you just cache the exception.

Comment (jzakaryan) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:624]: According to API reference, `commitSync` indeed throws a variety of exceptions other than `KafkaException`. However, I think there's a reason these exceptions are not handled here and are left to propagate up the call stack and I don't think there's a need to trigger the post-commit hook with the cached exception in this case.

**Review by somandal — APPROVED** [review]

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java
Comment (abhishekmendhekar) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java]: Assuming there is a set of actions that follow post commit using this hook.
true/false will give a signal of commit succeed or failed but an exception can help understand why it failed. Are you saying that there are more than 1 exception thrown? Generally Kafka should send only one commit message for a given consumer group and if this method commits more than 1 CG then I suggest returning a `Map<String, Exception>`. This gives user the most flexibility in terms of using this API.

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java
Comment (vmaheshw) [inline datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:627]: During shutdown, if it hits KafkaException, it returns true, which does not really mean success. This can impact the postCommit hook.

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java
Comment (vmaheshw) [inline datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:797]: Since, we don't have a concrete usage for this hook in open source and we currently have identified only the need for preCommit hook, will it be okay to postpone it, till either we identify the use for it or we get any pull/feature request. This will help us in avoiding an unnecessary tech debt to keep on maintaining.

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java
Comment (vmaheshw) [inline datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:612]: If preCommitHook() threw exception, the commit will not go through. Your current use-case to have audit related drain in the preCommitHook() also adds a dependency that if there is any failure in audit, the commits will not go through. Is this really expected ? Is it possible to have the audit system throwing errors while the regular producers are working fine.

**Review by vmaheshw — COMMENTED** [review]
Please evaluate

a)  if you really want to introduce the postCommitHook at this stage, where we don't have a clear usecase for it.
b) if you want to tie the audit failures to block the commits?

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java
Comment (jzakaryan) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:627]: This is no longer relevant now that post-commit hook is removed.

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java
Comment (jzakaryan) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:612]: I made it explicit in javadoc that the overriding class needs to take care of handling exceptions from pre-commit hook.

File: datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java
Comment (somandal) [inline datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java:613]: Can we remove this now that we no longer need to pass it to the postCommitHook()?

Comment (jzakaryan) [reply datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java]: Done

**Review by vmaheshw — APPROVED** [review]

**Review by somandal — APPROVED** [review]
Lgtm!

**Pull request merged**
